openwisp-monitoring/files/sbin/netjson-monitoring.lua

```diff
@@ -70,43 +70,52 @@ local host_interfaces = {}
 local dns_servers = {}
 local dns_search = {}
 
+local function get_wireless_netjson_interface(radio, name, iwinfo)
+  local clients = nil
+  local is_mesh = false
+  local htmode = radio.config.htmode
+  local netjson_interface = {
+    name = name,
+    type = 'wireless',
+    wireless = {
+      ssid = iwinfo.ssid,
+      mode = monitoring.wifi.iwinfo_modes[iwinfo.mode] or iwinfo.mode,
+      channel = iwinfo.channel,
+      frequency = iwinfo.frequency,
+      tx_power = iwinfo.txpower,
+      signal = iwinfo.signal,
+      noise = iwinfo.noise,
+      country = iwinfo.country,
+      htmode = htmode
+    }
+  }
+  -- if `iwinfo.channel` is empty then remove
+  -- the `wireless` table from the `netjson_interface`
+  if monitoring.utils.is_empty(iwinfo.channel) then
+    monitoring.utils.remove_table_key(netjson_interface, 'wireless')
+  end
```

Comment:
This solution is problematic, because the subsequent lines may try to add the clients to `netjson_interface.wireless` in edge cases, therefore in some cases this code may fail.

I think the solution here is to do something like the following (which is what I had originally requested): 

```lua
local netjson_interface = {
  name = name,
  type = 'wireless'
}
-- if channel is missing the WiFi interface is not fully up
-- and hence we avoid including its info because it will be rejected
if monitoring.utils.is_empty(iwinfo.channel) == false then
  wireless = {
    ssid = iwinfo.ssid,
    mode = monitoring.wifi.iwinfo_modes[iwinfo.mode] or iwinfo.mode,
    channel = iwinfo.channel,
    frequency = iwinfo.frequency,
    tx_power = iwinfo.txpower,
    signal = iwinfo.signal,
    noise = iwinfo.noise,
    country = iwinfo.country,
    htmode = htmode
  }
  if iwinfo.mode == 'Ad-Hoc' or iwinfo.mode == 'Mesh Point' then
    clients = ubus:call('iwinfo', 'assoclist', {device = name}).results
    is_mesh = true
  else
    local hostapd_output = ubus:call('hostapd.' .. name, 'get_clients', {})
    if hostapd_output then clients = hostapd_output.clients end
  end
  if not monitoring.utils.is_table_empty(clients) then
    netjson_interface.wireless.clients = monitoring.wifi.netjson_clients(clients,
      is_mesh)
  end
end
return netjson_interface
```

Reply:
I tried this approach initially but it failed many tests. Now I'm able to figure out where it's going wrong, thanks!